presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.LimitNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.google.common.collect.Range;
+
+import static com.facebook.presto.sql.planner.iterative.Lookup.noLookup;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public final class QueryCardinalityUtil
+{
+    private QueryCardinalityUtil()
+    {
+    }
+
+    public static boolean isScalar(PlanNode node)
+    {
+        return isScalar(node, noLookup());
+    }
+
+    public static boolean isScalar(PlanNode node, Lookup lookup)
+    {
+        return Range.singleton(1L).encloses(extractCardinality(node, lookup));
+    }
+
+    public static boolean isAtMostScalar(PlanNode node)
+    {
+        return isAtMostScalar(node, noLookup());
+    }
+
+    public static boolean isAtMostScalar(PlanNode node, Lookup lookup)
```

Comment:
Discussed with @martint, he doesn't have a better name right now, so we'll merge now and revisit.